### PR TITLE
Implement `ServerConfig`

### PR DESCRIPTION
**Pull request by guilload**
Implement `ServerConfig`.

### CODE REVIEW

File: quickwit-config/resources/tests/server_config/server.toml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+metastore_uri = "postgres://username:password@host:port/db"
+
+[indexer]
+data_dir_path = "/opt/quickwit/data"
```
Comment (fmassot) [inline quickwit-config/resources/tests/server_config/server.toml:4]: we can set up here `IndexingSplitStoreParams` (there 2 params)

File: quickwit-config/resources/tests/server_config/server.toml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+[searcher]
+data_dir_path = "/opt/quickwit/data"
+peer_seeds = [ "quickwit-searcher-0.local", "quickwit-searcher-1.local" ]
```
Comment (fmassot) [inline quickwit-config/resources/tests/server_config/server.toml:8]: no port?

File: quickwit-config/resources/tests/server_config/server.toml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+fast_field_cache_capacity = "1G"
+split_footer_cache_capacity = "500M"
+
+[storage]
+s3 = { region = "us-east-1", endpoint = "https://s3.us-east-1.amazonaws.com" }
```
Comment (fmassot) [inline quickwit-config/resources/tests/server_config/server.toml:13]: what is this s3 endpoint for?

File: quickwit-config/resources/tests/server_config/server.toml (hunk 1 above)
Comment (guilload) [reply quickwit-config/resources/tests/server_config/server.toml:4]: Will do.

File: quickwit-config/resources/tests/server_config/server.toml (hunk 2 above)
Comment (guilload) [reply quickwit-config/resources/tests/server_config/server.toml:8]: I guess not in this example :) We should be able to rely on default values, right?

This makes me think I need to set up the default listening ports in the server config.

File: quickwit-config/resources/tests/server_config/server.toml (hunk 3 above)
Comment (guilload) [reply quickwit-config/resources/tests/server_config/server.toml:13]: This is the URL of the AWS S3 API endpoint. Some accounts have custom endpoints (and the region is ignored).

File: quickwit-config/src/server_config.rs [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+use byte_unit::Byte;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct IndexerConfig {
+    pub data_dir_path: PathBuf,
+}
+
+impl IndexerConfig {
+    pub fn validate(&self) -> anyhow::Result<()> {
```
Comment (fmassot) [inline quickwit-config/src/server_config.rs:33]: maybe `check` instead of `validate` to be consistent

File: quickwit-config/src/server_config.rs [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        }
+        Ok(())
+    }
+}
+
+// TODO: caching
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct SearcherConfig {
+    pub data_dir_path: PathBuf,
+    pub peer_seeds: Vec<String>,
```
Comment (fmassot) [inline quickwit-config/src/server_config.rs:48]: why not `SocketAddr` instead of `String` for peer ?

File: quickwit-config/src/server_config.rs [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+impl SearcherConfig {
+    pub fn validate(&self) -> anyhow::Result<()> {
+        if !self.data_dir_path.exists() {
+            bail!(
+                "Data dir `{}` does not exist.",
+                self.data_dir_path.display()
+            );
+        }
+        if self.peer_seeds.is_empty() {
```
Comment (fmassot) [inline quickwit-config/src/server_config.rs:61]: mmm, this is ok to not have peer seeds.

File: quickwit-config/resources/tests/server_config/server.toml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+[indexer]
+data_dir_path = "/opt/quickwit/data"
+split_store_max_num_bytes = "100G"
+split_store_max_num_files = 1_000
+
+[searcher]
+data_dir_path = "/opt/quickwit/data"
+peer_seeds = [ "quickwit-searcher-0.local", "quickwit-searcher-1.local" ]
+fast_field_cache_capacity = "1G"
```
Comment (fulmicoton) [inline quickwit-config/resources/tests/server_config/server.toml:11]: there is also a `split_footer_cache_capacity`. We may call it `hotcache_cache_capacity` or `split_turbo_cache_capacity` any cool name you can think of :)

Comment (guilload) [reply quickwit-config/resources/tests/server_config/server.toml:11]: it's right below :D

**Comment by fulmicoton** [comment]
This is awesome work!

File: quickwit-config/src/server_config.rs (hunk 5 above)
Comment (guilload) [reply quickwit-config/src/server_config.rs:50]: `SocketAddr` only accepts IPs but we want to be able to provide hostnames.

File: quickwit-config/src/server_config.rs (hunk 4 above)
Comment (guilload) [reply quickwit-config/src/server_config.rs:38]: I find that `validate` comes up more frequently when it comes to validating configs or forms. Don't you agree?

**Review by fmassot — APPROVED** [review]

**Pull request merged**